File: android/src/main/kotlin/com/osim/healthkit/model/HuaWeiHealthKitSleepData.kt
package com.osim.healthkit.model

data class HuaWeiHealthKitSleepData (
    val store: MutableMap<String, Any?> = mutableMapOf()
) {
    constructor(
        type: Int ,
        startDate: Double,
        endDate: Double,
        metaData: String,
        deviceModel: String,
        deviceName: String,
        deviceType: String,
        deviceUniqueCode: String,
        updateTime: Double,
    ) : this() {
        this.type = type
        this.startDate = startDate
        this.endDate = endDate
        this.metaData = metaData
        this.deviceModel = deviceModel
        this.deviceName = deviceName
        this.deviceType = deviceType
        this.deviceUniqueCode = deviceUniqueCode
        this.updateTime = updateTime
    }

    var type: Int by store

    var startDate: Double by store

    var endDate: Double by store

    var metaData: String by store

    var deviceModel: String by store

    var deviceName: String by store

    var deviceType: String by store

    var deviceUniqueCode: String by store

    var updateTime: Double by store
}

File: lib/api/huawei_healthkit_sleep_data.dart
class HuaWeiHealthKitSleepData {
  int type;
  double startDate;
  double endDate;
  String metaData;
  String deviceModel;
  String deviceName;
  String deviceType;
  String deviceUniqueCode;
  double updateTime;

  HuaWeiHealthKitSleepData._(
    this.type,
    this.startDate,
    this.endDate,
    this.metaData,
    this.deviceModel,
    this.deviceName,
    this.deviceType,
    this.deviceUniqueCode,
    this.updateTime,
  );

  factory HuaWeiHealthKitSleepData.from(Map<Object?, Object?> rawData) =>
      HuaWeiHealthKitSleepData._(
        rawData['type'].castInt(-1),
        rawData['startDate'].castDouble(-1.0),
        rawData['endDate'].castDouble(-1.0),
        rawData['metaData'].castString(''),
        rawData['deviceModel'].castString(''),
        rawData['deviceName'].castString(''),
        rawData['deviceType'].castString(''),
        rawData['deviceUniqueCode'].castString(''),
        rawData['updateTime'].castDouble(-1.0),
      );

  @override
  String toString() {
    return 'HuaWeiHealthKitSleepData{type: $type, startDate: $startDate, endDate: $endDate, metaData: $metaData, deviceModel: $deviceModel, deviceName: $deviceName, deviceType: $deviceType, deviceUniqueCode: $deviceUniqueCode, updateTime: $updateTime}';
  }
}

extension on Object? {
  int castInt(int defVal) => _genericCast<int>(this, defVal);

  double castDouble(double defVal) => switch (runtimeType) {
        int => _genericCast<int>(this, defVal.toInt()).toDouble(),
        double => _genericCast<double>(this, defVal),
        _ => defVal,
      };

  String castString(String defVal) => _genericCast<String>(this, defVal);
}

T _genericCast<T>(dynamic val, T defVal) {
  if (val is T) return val;
  return defVal;
}
